client/rest/src/test/java/org/elasticsearch/client/RestClientSingleHostIntegTests.java

```diff
@@ -295,11 +295,10 @@ public void testRequestResetAndAbort() throws Exception {
                 httpGet.abort();
                 assertTrue(httpGet.isAborted());
                 try {
-                    assertTrue(future.isCancelled());
+                    assertTrue(future.isDone());
                     future.get();
-                    throw new AssertionError("exception should have been thrown");
                 } catch(CancellationException e) {
-                    //expected
+                    //expected sometimes - if the future was cancelled before executing successfully
```

Comment:
I tried to refresh my memory around this, and I am not sure what the best course of action is, but with this change I am no longer sure what this test is testing, because the future may be cancelled or not. I do see a problem with how the test was written and I am curious about why it took so long for this issue to surface: your explanation is evident, the request may be already completed because it gets cancelled.

Speculating, but this could also very well be a regression in the http client we use, as the future no longer gets cancelled, and if we relax the assertion the assumptions we rely upon no longer hold? 

I don't have great ideas about a solution, but these are my thoughts.

Reply:
Thanks for the details Luca. It took me a while to parse what the test was aiming to achieve too :) 

> I am no longer sure what this test is testing, because the future may be cancelled or not

That's a good point. I'd say this change makes sure the future completes successfully or cancelled, but I appreciate it might not be very useful (?) 

> if we relax the assertion the assumptions we rely upon no longer hold?

I think the test case is a bit overzealous though. I guess if the apache hhtp library breaks the invariant of the underlying future not being cancelled when the HTTP request is cancelled the above test case would catch it? Namely https://github.com/elastic/elasticsearch/pull/70308/files#diff-75a96c857e80e8d97c55e23176edffb572fa82587d5051907575e87798965e11R281-R288

The possibly flaky assertion I changed is testing a concurrency cancellation that IMO would even be hard for a user to verify - what is a correct behaviour? that the request was let to execute all the way or that it was absolutely always cancelled? (I'd argue that this actually depends on the request, when in the lifecycle of the request/response cycle the cancellation hits the client and the behaviour would be different in case of a long poll request for eg)

I don't have a strong opinion here, I'm happy to drop the test case altogether if you think that's better?

Reply:
@javanna sorry, I realised I was a bit dense before. What I meant was that given a user (or external observer) can't deterministically determine what a successful output is when an in progress request is cancelled (which are all the cases in a real life scenario) I think the only desireable and deterministically observable behaviour is that the request is not hanging - ie. the future is done/complete. 

So thinking about it some more, I think the modified assertion makes sense.

What do you think?

Reply:
Yes I think you are right. I was referring to the javadocs that I wrote, which is though not enough for me to remember exactly what I was worried about when I wrote this test :) I remember looking at the code of the async http client and being worried about the fact that the low level rest client reuses the same request object throughout multiple retries. We trust that the client sets the future reference correctly to the request object, so that the right future gets cancelled when the request gets aborted. I may have been overly paranoid here and wrote a test to test async http client behaviour more than that of our own client. I am happy with your recommendation, thanks for looking into this.